File: leetcode-everyday/二分查找/34-在排序数组中查找元素的第一个和最后一个位置/34-在排序数组中查找元素的第一个和最后一个位置.cpp
class Solution {
private:
    int lower_bound(vector<int>& nums, int target) {
        int left = 0, right = nums.size()-1;
        while(left <= right) {
            int mid = left + (right - left) / 2;
            if(nums[mid] >= target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
    int higher_bound(vector<int>& nums, int target) {
        int left = 0, right = nums.size()-1;
        while(left <= right) {
            int mid = left + (right - left) / 2;
            if(nums[mid] <= target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return right;
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {        
        int lower = lower_bound(nums, target);
        int higher = higher_bound(nums, target);
        if(lower > higher) {
            return vector<int>{-1, -1};
        }
        return vector<int>{lower, higher};
    }
};

File: leetcode-everyday/二分查找/34-在排序数组中查找元素的第一个和最后一个位置/34-在排序数组中查找元素的第一个和最后一个位置.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
 var searchRange = function(nums, target) {
  let lower = lower_bound(nums, target);
  let upper = upper_bound(nums, target);
  if(lower > upper) return [-1, -1];
  return [lower, upper];
};

function lower_bound(nums, target) {
  let left = 0;
  let right = nums.length - 1;
  while(left <= right) {
      let mid = left + Math.floor((right - left)/2);
      if(nums[mid] >= target) {
          right = mid - 1;
      } else {
          left = mid + 1;
      }
  }
  return left;
}

function upper_bound(nums, target) {
  let left = 0;
  let right = nums.length - 1;
  while(left <= right) {
      let mid = left + Math.floor((right - left)/2);
      if(nums[mid] <= target) {
          left = mid + 1;
      } else {
          right = mid - 1;
      }
  }
  return right;
}